### PR TITLE
EKF3: Add support for wheel encoder odometry

### PR DESCRIPTION
**Pull request by priseborough**
Adds the ability to use wheel encoder sensors on rovers to provide a low cost inertial odometry solution. Uses the exisiting body frame velocity observation model used by the ZED camera odometry interface.

Wheel axles are assumed to be parallel to the Y body axis.

Added as a placeholder to provide a cleanup of the EKF3 patches from https://github.com/ArduPilot/ardupilot/pull/6656

### CODE REVIEW

**Comment by priseborough** [comment]
@rmackay9 FYI

**Comment by rmackay9** [comment]
Merged, thanks!

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp [hunk 1]
```diff
@@ -1564,11 +1564,44 @@ void NavEKF3_core::SelectBodyOdomFusion()
...
+
+            // get the forward velocity
+            float fwdSpd = wheelOdmDataNew.delAng * wheelOdmDataNew.radius * (1.0f / wheelOdmDataNew.delTime);
+
+            // get the unit vector from the projection of the X axis onto the horizontal
+            Vector3f unitVec;
+            unitVec.x = prevTnb.a.x;
+            unitVec.y = prevTnb.a.y;
+            unitVec.z = 0.0f;
+            unitVec.normalized();
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp]: No, we want to get the horizontal component then normalise.

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp [hunk 2]
```diff
@@ -1564,11 +1564,44 @@ void NavEKF3_core::SelectBodyOdomFusion()
...
+
+            // multiply by forward speed to get velocity vector measured by wheel encoders
+            Vector3f velNED = unitVec * fwdSpd;
+
+            // This is a hack to enable use of the existing body frame velocity fusion method
+            // TODO write a dedicated observation model for wheel encoders
+            usingWheelSensors = true;
+            bodyOdmDataDelayed.vel = prevTnb * velNED;
+            bodyOdmDataDelayed.body_offset = wheelOdmDataNew.hub_offset;
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp]: yes, will fix

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp [hunk 3]
```diff
@@ -1564,11 +1564,44 @@ void NavEKF3_core::SelectBodyOdomFusion()
...
         // Fuse data into the main filter
         FuseBodyVel();
 
         // stop the performance timer
         hal.util->perf_end(_perf_FuseBodyOdom);
+
+    } else if (storedWheelOdm.recall(wheelOdmDataDelayed, imuDataDelayed.time_ms)) {
+
+        // check if the delta time is too small to calculate a velocity
+        if (wheelOdmDataNew.delTime > EKF_TARGET_DT) {
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp]: yes, will fix

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp [hunk 4]
```diff
@@ -1564,11 +1564,44 @@ void NavEKF3_core::SelectBodyOdomFusion()
...
+
+            // get the forward velocity
+            float fwdSpd = wheelOdmDataNew.delAng * wheelOdmDataNew.radius * (1.0f / wheelOdmDataNew.delTime);
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp]: yes, will fix

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp]: @pavloblindnology, nicely done!  hats off to you!

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (pavloblindnology) [reply libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp]: Yes, but normalizeD() doesn't normalize the vector itself - it's a function which returns normalized vector, which is not used in line #1588.
normalize() is a procedure which does normalize.

[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp]: OK, I understand what you mean. The way it is now will cause rovers with significant pitch angles to underestimate the distance travelled. Will fix.

[the remark this answers is not in the document]
Comment (pavloblindnology) [reply libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp]: Yes, exactly.

**Pull request closed without merging**